### PR TITLE
39 Add metrics class

### LINKED ISSUE #39: Add metadata output
## Feature
Add metadata output regardless of test fail or pass

### PR DESCRIPTION
**Pull request by Zejnilovic**
Will add unit tests for ComparisonResult by the end of the day. Not it is tested through the main class

Closes #39

### CODE REVIEW

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala [hunk 1]
```diff
@@ -75,18 +94,34 @@ object DatasetComparisonJob {
...
-        expectedDf.count(),
-        actualDf.count()
-      )
+    ComparisonResult(expectedDfRowCount, actualDfRowCount, 0, selector, resultDF, diffCount)
+  }
+
+  private def writeMetricsToFile(result: ComparisonResult, fileName: String): Unit = {
+    new File(fileName).mkdirs()
+    val writer = new PrintWriter(s"$fileName/_METRICS")
+//    val writer = new FileOutputStream(fileName);
```
Comment (HuvarVer) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala]: An extra line to delete.

File: datasetComparison/src/test/scala/za/co/absa/hermes/datasetComparison/ComparisonResultSuite.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package za.co.absa.hermes.datasetComparison
+
+import org.apache.spark.sql.Column
+import org.scalatest.FunSuite
+
+class ComparisonResultSuite extends FunSuite {
+
+  val CR1 = ComparisonResult(10, 11, 12, List.empty[Column], None, 13, "--alfa beta")
```
Comment (yruslan) [inline datasetComparison/src/test/scala/za/co/absa/hermes/datasetComparison/ComparisonResultSuite.scala]: Why `CR1` and `CR2` are outside the tests that check it?

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala (hunk 1 above)
Comment (yruslan) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala]: Redundant comment

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala [hunk 3]
```diff
@@ -40,7 +42,7 @@ object DatasetComparisonJob {
       )
       .getOrCreate()
 
-    execute(cliOptions)
+    execute(cliOptions, args.mkString(" "))
```
Comment (yruslan) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala:45]: You can do `args.mkString(" ")` inside `execute()` and save 1 argument.

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/ComparisonResult.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    import net.liftweb.json._
+    prettyRender(getMetadata)
+  }
+
+  def getMetadata: Map[String, String] = Map[String, String](
+      "referenceRowCount" -> refRowCount.toString,
+      "newRowCount" -> newRowCount.toString,
+      "numberOfDuplicates" -> duplicitiesCount.toString,
+      "passed" -> (diffCount == 0).toString,
+      "numberOfDifferences" -> diffCount.toString
```
Comment (yruslan) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/ComparisonResult.scala]: `passedOptions` is missing?

**Review by yruslan — COMMENTED** [review]
Looks good. Just some minor comments.

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/ComparisonResult.scala (hunk 4 above)
Comment (HuvarVer) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/ComparisonResult.scala]: Why is there no propagation of passedOptions?
I think there should be propagation.

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala [hunk 5]
```diff
@@ -49,21 +51,38 @@ object DatasetComparisonJob {
...
     * @param sparkSession Implicit spark session
     */
-  def execute(cliOptions: CliOptions)(implicit sparkSession: SparkSession): Unit = {
+  def execute(cliOptions: CliOptions, rawOptions: String = "")(implicit sparkSession: SparkSession): Unit = {
+    val result = compare(cliOptions)
+    val resultWithRawOpts = result.copy(passedOptions = rawOptions)
+
+    if (resultWithRawOpts.resultDF.isDefined) {
+      resultWithRawOpts.resultDF.get.write.format("parquet").save(cliOptions.outPath)
+    }
```
Comment (GeorgiChochov) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala]:
```suggestion
    resultWithRawOpts.resultDF.foreach { df =>
      df.write.format("parquet").save(cliOptions.outPath)
    }
```

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala [hunk 6]
```diff
@@ -157,7 +192,7 @@ object DatasetComparisonJob {
   private def createDiffDataFrame(keys: Seq[String],
                                           path: String,
                                           expectedMinusActual: DataFrame,
-                                          actualMinusExpected: DataFrame): Unit = {
+                                          actualMinusExpected: DataFrame): DataFrame = {
```
Comment (GeorgiChochov) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala]: Alignment seems off here

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala [hunk 7]
```diff
@@ -75,18 +94,34 @@ object DatasetComparisonJob {
...
 
-    if ((expectedExceptActual.count() + actualExceptExpected.count()) == 0) {
-      scribe.info("Expected and actual data sets are the same.")
-    } else {
-      createDiffDataFrame(keys, cliOptions.outPath, expectedExceptActual, actualExceptExpected)
+    val resultDF: Option[DataFrame] = (expectedExceptActual.count() + actualExceptExpected.count()) match {
+      case 0 => None
+      case _ => Some(createDiffDataFrame(keys, cliOptions.outPath, expectedExceptActual, actualExceptExpected))
+    }
+    val diffCount: Long = if (resultDF.isDefined) resultDF.get.count() else 0
```
Comment (GeorgiChochov) [inline datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala]:
```suggestion
    val diffCount: Long = resultDF.map(_.count).getOrElse(0)
```

File: datasetComparison/src/test/scala/za/co/absa/hermes/datasetComparison/ComparisonResultSuite.scala (hunk 2 above)
Comment (Zejnilovic) [reply datasetComparison/src/test/scala/za/co/absa/hermes/datasetComparison/ComparisonResultSuite.scala]: Had bigger plans, settled for less.

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/ComparisonResult.scala (hunk 4 above)
Comment (Zejnilovic) [reply datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/ComparisonResult.scala]: Will add

Comment (Zejnilovic) [reply datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/ComparisonResult.scala]: Will add

File: datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala (hunk 3 above)
Comment (yruslan) [reply datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparisonJob.scala:45]: 😄 My bad, confused `cliOptions` and `args`.

**Comment by Zejnilovic** [comment]
Build passes again, sorry for that

**Review by GeorgiChochov — CHANGES_REQUESTED** [review]

**Review by GeorgiChochov — APPROVED** [review]

**Review by HuvarVer — APPROVED** [review]

**Pull request merged**